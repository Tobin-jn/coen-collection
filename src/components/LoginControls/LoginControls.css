.LoginControls {
  background-color: #e8e8e8;
  width: 100%
}


.login-form {
  display: flex;
  height: 76px;
  justify-content: space-evenly;
  margin: auto;
  width: 60%
}

input {
  border: 4px solid black;
  border-radius: 30px;
  bottom: 10px;
  font-size: 22px;
  height: 50px;
  margin: 10px 0 ;
  padding-left: 15px;
  position: relative;
}

.submit-btn {
  background: lightgrey;
  border: 4px solid black;
  border-radius: 30px;
  bottom: 10px;
  color: black;
  font-size: 24px;
  font-weight: bolder;
  margin: 10px 0px;
  padding: 10px;
  position: relative;
  text-align: center;
  text-decoration: none;
  width: 250px;
}

.submit-btn:hover {
  background: grey;
  color: lightgrey;
  transition: all 0.4s ease-in-out;
}