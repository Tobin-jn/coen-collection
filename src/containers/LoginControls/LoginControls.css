.LoginControls {
  background-color: #e8e8e8;
  height: 56px;
}

.login-form {
  display: flex;
  height: 56px;
  justify-content: space-evenly;
  margin: auto;
  width: 80%;
}

input {
  border: 4px solid black;
  border-radius: 30px;
  bottom: 10px;
  font-size: 22px;
  height: 28px;
  margin: 10px 0;
  padding-left: 15px;
  position: relative;
}

.submit-btn {
  background: lightgrey;
  border: 4px solid black;
  border-radius: 30px;
  bottom: 10px;
  color: black;
  cursor: pointer;
  font-size: 19px;
  font-weight: bolder;
  margin: 10px 0px;
  padding: 4px;
  position: relative;
  text-align: center;
  text-decoration: none;
  width: 250px;
}

.submit-btn:hover {
  background: grey;
  color: lightgrey;
  transition: all 0.4s ease-in-out;
}

.error-msg {
  bottom: 20px;
  color: red;
  font-size: 66;
  position: relative;
  text-align: center;
}
