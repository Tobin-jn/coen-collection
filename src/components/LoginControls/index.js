import React, { Component } from "react";
import PropTypes from "prop-types";
import "./LoginControls.css";
import { logIn, saveName } from "../../actions";
import { connect } from "react-redux";
import { Link, Redirect } from "react-router-dom";



class LoginControls extends Component {
  constructor(props) {
    super(props);
    this.state = {
      email: "",
      password: "",
      name: "",
      pathname: this.props.location.pathname,
      error: false,
      errorMessage: ""
    };
  }

  handleInputChange = e => {
    const { name, value } = e.target;
    this.setState({ [name]: value });
  };

  handleSubmit = async e => {
    e.preventDefault();
    if (this.state.pathname === "/login") {
      this.loginUser();
    } else {
      this.signupUser();
    }
  };

  validateEmail = () => {
    return !this.state.email.includes("@");
  };

  async loginUser() {
    if (this.validateEmail()) {
      this.setState({ errorMessage: "Please enter a valid e-mail address" });
      return;
    }
    try {
      const response = await fetch("http://localhost:3000/api/users", {
        method: "POST",
        credentials: "same-origin",
        body: JSON.stringify({
          email: this.state.email,
          password: this.state.password
        }),
        headers: { "Content-Type": "application/json" }
      });
      const data = await response.json();

      this.props.saveName(data.data.name)
      this.props.handleLogin(true)

    } catch (error) {
      this.setState({ errorMessage: "Invalid e-mail/password" });
    }
  }

  async signupUser() {
    if (this.validateEmail()) {
      this.setState({ errorMessage: "Please enter a valid e-mail address" });
      return;
    }
    let data;
    const response = await fetch("http://localhost:3000/api/users/new", {
      method: "POST",
      body: JSON.stringify({
        name: this.state.name,
        email: this.state.email,
        password: this.state.password
      }),
      headers: { "Content-Type": "application/json" }
    });
    data = await response.json();

    console.log(data)
    this.props.saveName(this.state.name)
    this.props.handleLogin(true)
    // this.getUserName()

    if (data.error && data.error.includes("already exists")) {
      this.setState({ errorMessage: "User account already exists!" });
    }
    //this.setState = { email: "", password: "", name: "" };
  }


  render() {
    return (
        <div className="LoginControls">
          {!this.props.loggedIn &&
            <form className="login-form" onSubmit={this.handleSubmit}>
              <input
                name="email"
                placeholder="email"
                className="email"
                value={this.state.email}
                onChange={this.handleInputChange}
              />
              {this.state.pathname === "/signup" && (
                <input
                  name="name"
                  placeholder="name"
                  className="name"
                  value={this.state.name}
                  onChange={this.handleInputChange}
                />
              )}
              <input
                name="password"
                placeholder="password"
                className="password"
                value={this.state.password}
                onChange={this.handleInputChange}
              />
              <button
                className="submit-btn"
                disabled={!this.state.email || !this.state.password}
              >submit
              </button>
              <p className="error-msg">{this.state.errorMessage}</p>
            </form>
          }

          { this.props.loggedIn && 
            <div className="loggedIn">
              <Redirect to='/' />
            </div>
          }

        </div>
    );
  }
}

const mapStateToProps = (state) => ({
  loggedIn: state.loggedIn
})

const mapDispatchToState = (dispatch) => ({
  handleLogin: (loggedIn) => dispatch(logIn(loggedIn)),
  saveName: (name) => dispatch(saveName(name)) 
})

export default connect(mapStateToProps, mapDispatchToState)(LoginControls);
