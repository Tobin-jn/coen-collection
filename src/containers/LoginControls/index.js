import React, { Component } from "react";
import PropTypes from "prop-types";
import "./LoginControls.css";
import {
  logIn,
  saveUserData,
  toggleFavorite,
  setErrorMessage
} from "../../actions";
import { connect } from "react-redux";
import { Redirect } from "react-router-dom";
import * as fetch from "../../utilities/fetch.js";

export class LoginControls extends Component {
  constructor(props) {
    super(props);
    this.state = {
      email: "",
      password: "",
      username: "",
      pathname: this.props.location.pathname,
      error: false
    };
  }

  handleInputChange = e => {
    const { name, value } = e.target;
    this.setState({ [name]: value });
  };

  handleSubmit = e => {
    e.preventDefault();
    this.clearErrorMessage();
    if (this.validateEmail() && this.validateInputLength("password", 6)) {
      this.state.pathname === "/login" ? this.loginUser() : this.signupUser();
    }
  };

  validateEmail = () => {
    var re = /^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    if (re.test(this.state.email)) {
      return true;
    }
    this.props.handleErrorMessage("Please enter a valid e-mail address");
    return false;
  };

  validateInputLength = (inputType, length) => {
    if (this.state[inputType].length >= length) {
      return true;
    } else {
      this.props.handleErrorMessage(
        `${inputType} must be at least ${length} characters`
      );
    }
  };

  loginUser = async () => {
    try {
      const { email, password } = this.state;
      const fetchUser = await fetch.fetchLoginUser(email, password);

      this.updateUserDataAfterLogin(fetchUser)

    } catch (error) {
      this.props.handleErrorMessage("Invalid e-mail/password");
    }
  };

  updateUserDataAfterLogin = (userData) => {
    this.props.saveUserData(userData.data.name, userData.data.id);
    this.props.handleLogin(true);
    this.getUserFavorites(userData.data.id);
  }

  getUserFavorites = async userId => {
    try {
      const fetchFavorites = await fetch.retrieveUserFavorites(userId);
      fetchFavorites.forEach(fav =>
        this.props.handleFavoriteToggle(fav.movie_id)
      );
    } catch (error) {
      this.props.handleErrorMessage("Favorites error")
    }
  };

  signupUser = async () => {
    if (!this.validateInputLength("username", 2)) {
      return;
    }
    const { username, email, password } = this.state;
    const fetchSignup = await fetch.fetchSignupUser(username, email, password);

    if (fetchSignup.error && fetchSignup.error.includes("already exists")) {
      this.props.handleErrorMessage("User account already exists!");
    } else {
      this.updateUserDataAfterSignup(username, fetchSignup.id)
    }
  };

  updateUserDataAfterSignup = (username, id) => {
    this.props.saveUserData(username, id);
    this.props.handleLogin(true);
  }

  clearErrorMessage = () => {
    this.props.handleErrorMessage("");
  };

  render() {
    return (
      <div className="LoginControls">
        {!this.props.loggedIn && (
          <form className="login-form" onSubmit={this.handleSubmit}>
            <input
              name="email"
              placeholder="email"
              value={this.state.email}
              className='email'
              onChange={this.handleInputChange}
            />
            {this.state.pathname === "/signup" && (
              <input
                name="username"
                placeholder="username"
                className='username'
                value={this.state.username}
                onChange={this.handleInputChange}
              />
            )}
            <input
              name="password"
              placeholder="password"
              className="password"
              value={this.state.password}
              onChange={this.handleInputChange}
            />
            <button className="submit-btn">submit</button>
          </form>
        )}
        {this.props.loggedIn && (
          <div className="loggedIn">
            <Redirect to="/" />
          </div>
        )}
        <section>
          <p className="error-msg">{this.props.errorMessage}</p>
        </section>
      </div>
    );
  }
}

export const mapStateToProps = state => ({
  loggedIn: state.loggedIn,
  userId: state.currentUser.id,
  errorMessage: state.errorMessage
});

export const mapDispatchToProps = dispatch => ({
  saveUserData: (username, id) => dispatch(saveUserData(username, id)),
  handleLogin: loggedIn => dispatch(logIn(loggedIn)),
  handleErrorMessage: message => dispatch(setErrorMessage(message)),
  handleFavoriteToggle: id => dispatch(toggleFavorite(id))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(LoginControls);
