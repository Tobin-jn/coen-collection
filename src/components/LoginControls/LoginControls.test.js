import React from "react";
import { shallow } from "enzyme";
import LoginControls from "./index";

describe("Card", () => {
  it("Renders like snapshot", () => {
    const wrapper = shallow(<LoginControls />)
    expect(wrapper).toMatchSnapshot();
  });
});