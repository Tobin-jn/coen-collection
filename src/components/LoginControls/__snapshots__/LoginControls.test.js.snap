// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Renders like snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LoginControls />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          className="user-name"
          name="username"
          onChange={[Function]}
          placeholder="username"
          value=""
        />,
        <input
          className="password"
          name="password"
          onChange={[Function]}
          placeholder="password"
          value=""
        />,
        <button
          disabled={true}
        >
          submit
        </button>,
      ],
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "user-name",
          "name": "username",
          "onChange": [Function],
          "placeholder": "username",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "password",
          "name": "password",
          "onChange": [Function],
          "placeholder": "password",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "submit",
          "disabled": true,
        },
        "ref": null,
        "rendered": "submit",
        "type": "button",
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            className="user-name"
            name="username"
            onChange={[Function]}
            placeholder="username"
            value=""
          />,
          <input
            className="password"
            name="password"
            onChange={[Function]}
            placeholder="password"
            value=""
          />,
          <button
            disabled={true}
          >
            submit
          </button>,
        ],
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "user-name",
            "name": "username",
            "onChange": [Function],
            "placeholder": "username",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "password",
            "name": "password",
            "onChange": [Function],
            "placeholder": "password",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "submit",
            "disabled": true,
          },
          "ref": null,
          "rendered": "submit",
          "type": "button",
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
